CONTRIBUTING.md

```diff
@@ -67,6 +67,89 @@ Some characteristics of good tests:
 * be minimal, deterministic, stable (unaffected by irrelevant changes), easy to understand and review
 * have minimal dependencies: a compiler bug test should not depend on, e.g., the Scala library
 
+There are mainly three kinds of tests: unit tests, property-based tests, and integration tests.
+
+#### JUnit
+
+For unit tests we use JUnit, which can be run from sbt shell as follows:
+
+```
+root> junit/testQuick
+```
+
+It might take a few minutes the first time you run `junit/testQuick`, but from the second time onwards
+sbt will only run the tests that is affected by the code change since the last run.
+See `test/junit/` for the examples of JUnit tests.
+
+#### ScalaCheck
+
+For testing that can benefit from having lots of randomly generated data, property-based testing should be used.
+This is run from sbt shell as follows:
+
+```
+root> scalacheck/testOnly ByOneRangeTest
+```
+
+See `test/scalacheck/range.scala`.
+
+#### Partest
+
+scala/scala comes with a powerful integration testing tool called Partest.
+Using Partest you can compile or run some Scala code, and compare it against the expected output.
+In the source tree, partests are located under `test/files/<category>/`. The main categories are:
+
+- `pos`: These files must compile successfully.
+- `run`: In addition to compiling, `Test.main` is run and its output is compared against the test's `.check` file.
+- `neg`: These files must NOT compile, with compiler output matching the expected output in the `.check` file.
+- Other categories such as `jvm` behave the same as `run` category.
+
+To run a single negative test from sbt shell:
+
+```
+root> test/it:testOnly partest -- test/files/neg/delayed-init-ref.scala
```

Comment:
Surely just `root> partest test/files/neg/delayed-init-ref.scala` would do the same job here?

Reply:
Also worth mentioning `--verbose` for single tests ... IME when running a single test you almost always want `--verbose`.

Reply:
I am a noob contributor to scala/scala, so I am writing this as I am figuring out :) I'll change to `partest` command. More tips and tricks welcome from you and others.